tests/dispatch/test_connection_pools.py

```diff
@@ -131,3 +131,35 @@
         await response.close()
         assert len(http.active_connections) == 0
         assert len(http.keepalive_connections) == 0
+
+
+@pytest.mark.asyncio
+async def test_keepalive_connection_closed_by_server_is_reestablished(server):
+    """
+    Upon keep-alive connection closed by remote a new connection should be reestablished.
+    """
+    async with httpx.ConnectionPool() as http:
+        response = await http.request("GET", "http://127.0.0.1:8000/")
+        await response.read()
+
+        await server.shutdown()  # shutdown the server to close the keep-alive connection
+        await server.startup()
+
+        response = await http.request("GET", "http://127.0.0.1:8000/")
+        await response.read()
+        assert len(http.active_connections) == 0
+        assert len(http.keepalive_connections) == 1
+
+
+@pytest.mark.asyncio
+async def test_connection_closed_by_server_before_receiving_data(server):
+    """
+    If the server closes the connection after sending the response but before reading the data
+    a disconnect error is raised.
+    """
+    async with httpx.ConnectionPool() as http:
+        response = await http.request("GET", "http://127.0.0.1:8000/")
+        await server.shutdown()  # shutdown the server to close the connection before receiving the data
+
+        with pytest.raises(httpx.exceptions.NotConnected):
```

Comment:
Is `NotConnected` the right exception here? Maybe we should wrap it into a more descriptive Exception and error message?